src/hotspot/share/gc/parallel/psCompactionManager.cpp

```diff
@@ -115,42 +119,46 @@ ParCompactionManager::gc_thread_compaction_manager(uint index) {
   return _manager_array[index];
 }
 
-inline void ParCompactionManager::publish_and_drain_oop_tasks() {
-  oop obj;
-  while (oop_stack()->pop_overflow(obj)) {
-    if (!oop_stack()->try_push_to_taskqueue(obj)) {
-      follow_contents(obj);
-    }
-  }
-  while (oop_stack()->pop_local(obj)) {
-    follow_contents(obj);
-  }
+void ParCompactionManager::push_objArray(oop obj) {
+  assert(obj->is_objArray(), "precondition");
+  _mark_and_push_closure.do_klass(obj->klass());
+
+  objArrayOop obj_array = objArrayOop(obj);
+  size_t array_length = obj_array->length();
+  size_t initial_chunk_size =
+    // The destination array is unused when processing states.
+    _partial_array_splitter.start(&_marking_stack, obj_array, nullptr, array_length);
+  int end = checked_cast<int>(initial_chunk_size);
```

Comment:
I wonder if we can change the callee so that we don't need `checked_cast` here. (Intuitively, [start, end) denotes a range and both number should never be negative.)